ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-composite-answer/activityCompositeAnswer.component.ts

```diff
@@ -42,36 +78,80 @@ export class ActivityCompositeAnswer implements OnChanges {
-                    // Important: if a user answered only on the last question, backend returns the correct array of answers
-                    this.childQuestionBlocks = questionsRows.map((currentRow: ActivityQuestionBlock<any>[]) => {
-                        if (currentRow.length !== blankRow.length) {
-                            const missedQuestions = blankRow.slice(currentRow.length);
-                            currentRow.push(...missedQuestions);
-                        }
-                        return currentRow;
-                    });
+                    this.childQuestionBlocks = this.rebuildChildQuestions(newBlock, newAnswers);
+                }
+            }
+        }
+    }
+
+    private updateEquationQuestions(equations: AnswerResponseEquation[] = []): void {
+        // will keep here equations values which were updated
+        const equationsToUpdate: {[stableId: string]: DecimalAnswer[]} = {};
+
+        const childEquations = this.block.children
+            .filter(question => question.questionType === QuestionType.Equation) as  ActivityEquationQuestionBlock[];
+
+        // add equations answers to the parent composite question answer
+        for (const equation of equations) {
+            for (const childEquation of childEquations) {
+                if (childEquation.stableId === equation.stableId) {
+                    equationsToUpdate[childEquation.stableId] = equation.values;
+
+                    const prevAnswer: AnswerContainer[][] = this.block.answer;
+                    const newAnswer: AnswerContainer[][] = prevAnswer.map((answerRow: AnswerContainer[], index) =>
+                        // eslint-disable-next-line arrow-body-style
+                        answerRow.map((answer: AnswerContainer) => {
+                            return (answer.stableId === childEquation.stableId) ? {
+                                ...answer,
+                                value: [equation.values[index]]
+                            } : answer;
+                        })
+                    );
+                    this.block.setAnswer(newAnswer, false);
                 }
             }
         }
+
+        // update equation questions answers (inside of the parent composite question)
+        // which values were changed (get them from `equationsToUpdate`)
+        // in order to re-render them only
+        this.childAnswerComponents.toArray()
+            .filter((answer: ActivityAnswerComponent) => Object.keys(equationsToUpdate).includes(answer.block.stableId))
+            .map((answer: ActivityAnswerComponent) => answer.equationAnswerComponent)
+            .forEach((equationAnswerComponent: ActivityEquationAnswerComponent, index: number) => {
+                const newValue = [equationsToUpdate[equationAnswerComponent.block.stableId][index]];
+                equationAnswerComponent.block.setAnswer(newValue);
+            });
+    }
```

Comment:
Are we sure that for composite we need to do all this? The child equation questions are listening for their changes, so this code and the code in the equation component are both updating? Or you just looking in the equation component for answers of size 1? If so maybe we should add a comment?

Reply:
Removed that, see https://github.com/broadinstitute/ddp-angular/pull/1222#discussion_r866772902